engine/src/main/java/org/terasology/engine/world/generation/WorldBuilder.java

```diff
@@ -200,92 +209,69 @@ public World build() {
         return result;
     }
 
-    private void determineProviderChainFor(Class<? extends WorldFacet> facet,
-                                           ListMultimap<Class<? extends WorldFacet>, FacetProvider> result,
-                                           boolean scalable) {
-        if (result.containsKey(facet)) {
-            return;
-        }
-        if (!facetCalculationInProgress.add(facet)) {
-            throw new RuntimeException("Circular dependency detected when calculating facet provider ordering for " + facet);
-        }
-        Set<FacetProvider> orderedProviders = Sets.newLinkedHashSet();
-
-        // first add all @Produces facet providers
+    private void addProviderChain(Class<? extends WorldFacet> facet, boolean scalable, int minPriority,
+                                  Set<FacetProvider> orderedProviders) {
```

Comment:
Could you please provide some Javadoc what this method is doing (could also be a general follow-up PR to add docstrings if you prefer). I few things I noticed from just looking at the PR

- `orderedProviders` is a "destination" parameter and is populated in this method
- `providedBy` is updated
- there is a "hidden" recursion via `addRequirements`